ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java

```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
     }
     ThrottlingRecommendation throttlingRecommendation;
-    Timer.Context timer = quotaMetrics.quotaEnforcementTime.time();
+    Timer.Context timer = quotaMetrics.quotaRecommendationTime.time();
     try {
       List<QuotaRecommendation> quotaRecommendations = new ArrayList<>();
-      for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
+      boolean partialRecommendation = false;
+      boolean noRecommendation = true;
+      for (QuotaEnforcer quotaEnforcer : quotaEnforcers) {
         try {
           quotaRecommendations.add(quotaEnforcer.recommend(restRequest));
+          noRecommendation = false;
         } catch (QuotaException quotaException) {
-          logger.warn("Could not get recommendation for quota {} due to exception: {}",
+          LOGGER.warn("Could not get recommendation for quota {} due to exception: {}",
               quotaEnforcer.supportedQuotaNames(), quotaException.getMessage());
+          partialRecommendation = true;
         }
       }
-      if (quotaRecommendations.isEmpty()) {
-        quotaMetrics.quotaNotEnforcedCount.inc();
+      if (noRecommendation) {
+        quotaMetrics.noRecommendationRate.mark();
+      } else if (partialRecommendation) {
+        quotaMetrics.partialQuotaRecommendationRate.mark();
       }
       throttlingRecommendation = quotaRecommendationMergePolicy.mergeEnforcementRecommendations(quotaRecommendations);
       if (throttlingRecommendation.shouldThrottle()) {
-        quotaMetrics.quotaExceededCount.inc();
+        quotaMetrics.quotaExceedRecommendationRate.mark();
       }
     } finally {
       timer.stop();
     }
     return throttlingRecommendation;
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * This implementation does check and charge in two phases.
+   * In the first phase, it obtains recommendations from all the {@link QuotaEnforcer}s and uses the recommendations (in
+   * conjunction with the parameters) to determine if the request should be charged. Any exception from any
+   * {@link QuotaEnforcer}s in this phase are bubbled up to the caller and no charge is made.
+   *
+   * In the second phase, after it has determined that the charge needs to be done, it attempts to charge the individual
+   * {@link QuotaEnforcer}s one by one. The behavior in this phase is to make a best effort to charge as many
+   * {@link QuotaEnforcer}s as possible and send a {@link QuotaAction#ALLOW} recommendation. If there is an exception in
+   * any {@link QuotaEnforcer#charge} call, that exception is logged and charge is still attempted for rest of the
+   * {@link QuotaEnforcer}s. As such, there is a potential for partial charge to happen in this implementation.
+   */
   @Override
-  public ThrottlingRecommendation charge(RestRequest restRequest, BlobInfo blobInfo,
-      Map<QuotaName, Double> requestCostMap) {
-    if (requestQuotaEnforcers.isEmpty()) {
-      return null;
+  public QuotaAction chargeAndRecommend(RestRequest restRequest, BlobInfo blobInfo,
```

Comment:
doesn't seem like that BlobInfo is used anywhere, now that we are changing the interface, we might as well just remove the BlobInfo once for all.

Reply:
done.